### ISSUE TITLE
Within-Field Docstring Merging

### ISSUE THREAD

**Issue by Zeitsperre** [rsokl/custom_inherit#25]
I'm wondering if there is any way to handle a case where parameters or comments are composed from multiple sections.

It's always easier to understand with an example:
```
def some_operation(func):
    @doc_inherit(parent=func, style="numpy_napoleon")
    def wrapper( ... ):
        """
        Parameters
        ---------------
        x : str
          some string
        y : str
          some other string
        """
        < stuff happens >
        return
    return wrapper

@some_operation
def foo( ... ):
    """
    Parameters
    ---------------
    z : int
      some number
    """
    < stuff happens >
```
Then on inspection we get something resembling:
```
    Parameters
    ---------------
    x : str
      some string
    y : str
      some other string
    z : int
      some number
```
I can understand how technically challenging this could be but I'm wondering if this is a desired feature. Does this seem feasible?

**Comment by rsokl** [comment]
Hi @Zeitsperre thanks for the nice example.

This is certainly feasible - I wrote a merge-style like this a few years ago, and it wasn't all that bad to implement (unfortunately there is no trace of it...). That being said, you could certainly leverage the [docstring parsing tools](https://github.com/rsokl/custom_inherit/blob/master/src/custom_inherit/_doc_parse_tools/napoleon_parse_tools.py) that are included in this library to implement this relatively easily. 

Specifically `parse_napoleon_doc`  will give you a dictionary containing all of the sections of the parsed docstring. You can access the `Parameters` section of this and parse it for the individual parameters within and do a specialized merge for this.

Your use case with a wrapper is an interesting one that I hadn't considered. I would definitely be open to adding this strategy to the zoo.

**Comment by Zeitsperre** [comment]
Ok sounds pretty cool. I'll see when I have time to work on this but I'll start by forking master and examining the algorithm based off your test examples. What would be good to know going in is how best to handle the order of signature elements (ie: parent args --> child args --> parent kwargs --> child kwargs).

Is there currently a way to link the entries under Parameters to their call signatures to know which is an arg/kwarg?

**Comment by rsokl** [comment]
Presently, is no way to do that. I would be open to making these call-signatures available to the merge styles in version `3.0` after dropping support for Python 2.7. 

So that I have more context: in what case would you want to know if an argument is an arg/kwarg, for the purpose of docstring-merging?

**Comment by Boubsi** [comment]
Hi all,

I am also very interested in such a merge function, and I developed three new merge styles : numpy_with_merge, google_with_merge and numpy_napoleon_with_merge, where section content are also merged. 

Before I put forward a merge request, i was just wondering : how do you run all the tests since your tests modules does not inherit from unitest ?

Please tell me so I can test all the module.

**Comment by Boubsi** [comment]
Hi all,

I am also very interested in such a option, and I developed three new built-in styles that allow such inner section merging.  You can find it on my repo https://github.com/Boubsi/custom_inherit/tree/inner_section_merging.

The new built-in styles are called (for now) `"numpy_with_merge"`, `"numpy_napoleon_with_merge"` and `"google_with_merge"`.

With those styles, if there is an overlap in some section of child and parents docstrings, the resulting section will be the concatenation of parent's and child's section. It works for all sections except some sections for which it is not relevant to perform the merge (such as "Summary" or "Examples" sections. 

For example, here is the docstring I wrote for the `"numpy_with_merge"` style:
```
    """
    Behaves identically to the 'numpy' style, but also merges sections that
    overlap, instead of only keeping the child's section. All sections are
    concerned except sections "Short Summary", "Extended Summary", "Deprecation
    Warning" and "Examples" for which the 'numpy' style behaviour applies.

    Example:
        - parent's docstring:

            ''' Parent's line

                Parameters
                ----------
                x: int
                    description of x
                y: Union[None, int]
                    description of y

                Raises
                ------
                NotImplemented Error

                Example
                -------
                >>> parent_func(x=3, y=None)
                NotImplementedError:'''

        - child's docstring:

            ''' Child's line

                Parameters
                ----------
                z: Union[None, int]
                    description of z

                Returns
                -------
                int

                Notes
                -----
                notes blah blah

                Example
                -------
                >>> child_func(x=3, y=None, z=4)
                7'''

        - docstring that is ultimately inherited:

            ''' Child's line

                Parameters
                ----------
                x: int
                    description of x
                y: Union[None, int]
                    description of y
                z: Union[None, int]
                    description of z

                Returns
                -------
                int

                Notes
                -----
                notes blah blah

                Example
                -------
                >>> child_func(x=3, y=None, z=4)
                7'''
```

What do you think about it? Should I make a merge request on this repo?

**Comment by rsokl** [comment]
@Boubsi this looks like a nice addition to the styles. I think that it would be a convenient add-on for users who want this sort of capability. 

A PR would be welcome!

**Comment by rsokl** [comment]
Closed by #31
